clients/src/main/java/org/apache/kafka/clients/consumer/internals/NetworkClientDelegate.java

```diff
@@ -0,0 +1,287 @@
+            }
+
+            if (!doSend(unsent, currentTimeMs, unsentAndUnreadyRequests)) {
+                log.debug("No broker available to send the request: {}", unsent);
+                unsent.callback.ifPresent(v -> v.onFailure(Errors.NETWORK_EXCEPTION.exception(
+                        "No node available in the kafka cluster to send the request")));
+            }
+        }
+
+        if (!unsentAndUnreadyRequests.isEmpty()) {
+            // Handle the unready requests in the next event loop
+            unsentRequests.addAll(unsentAndUnreadyRequests);
+        }
+    }
+
+    // Visible for testing
+    boolean doSend(final UnsentRequest r,
+                   final long currentTimeMs,
+                   final Queue<UnsentRequest> unsentAndUnreadyRequests) {
+        Node node = r.node.orElse(client.leastLoadedNode(currentTimeMs));
+        if (node == null || nodeUnavailable(node)) {
+            return false;
+        }
+        ClientRequest request = makeClientRequest(r, node);
+        if (client.isReady(node, currentTimeMs)) {
+            activeNodes.add(node);
+            client.send(request, currentTimeMs);
+        } else {
+            // enqueue the request again if the node isn't ready yet. The request will be handled in the next iteration
+            // of the event loop
+            log.debug("Node is not ready, handle the request in the next event loop: node={}, request={}", node, r);
+            unsentAndUnreadyRequests.add(r);
+        }
+        return true;
+    }
+
+    private void checkDisconnects() {
+        // Check the connection status by all the nodes that are active. Disconnect the disconnected node if it is
+        // unable to be connected.
+        Iterator<Node> iter = activeNodes.iterator();
+        while (iter.hasNext()) {
+            Node node = iter.next();
+            iter.remove();
+            if (client.connectionFailed(node)) {
+                client.disconnect(node.idString());
+            }
+        }
+    }
+
+    private ClientRequest makeClientRequest(final UnsentRequest unsent, final Node node) {
+        return client.newClientRequest(
+                node.idString(),
+                unsent.abstractBuilder,
+                time.milliseconds(),
+                true,
+                (int) unsent.timer.remainingMs(),
+                unsent.callback.orElse(new DefaultRequestFutureCompletionHandler()));
+    }
+
+    public void maybeTriggerWakeup() {
+        if (this.wakeup) {
+            this.wakeup = false;
+            throw new WakeupException();
+        }
+    }
+
+    public void wakeup() {
+        this.wakeup = true;
+        this.client.wakeup();
+    }
+
+    public Node leastLoadedNode() {
+        return this.client.leastLoadedNode(time.milliseconds());
+    }
+
+    public void add(final UnsentRequest r) {
+        r.setTimer(this.time, this.requestTimeoutMs);
+        unsentRequests.add(r);
+    }
+
+    public void ready(final Node node) {
```

Comment:
nit: we should resist exposing all of these if we can. I think we want a really basic api.